CIP-0014/CIP-0014.md

```diff
@@ -0,0 +1,116 @@
+policy id nor the asset name are intended to be human-readable data. 
+
+On the one hand, the policy id is a hash digest of either a monetary script or a Plutus script. On the other hand, the asset name is an arbitrary bytestring of up to 32 bytes (which does not necessarily decode to a valid UTF-8 sequence). In addition, it is possible for an asset to have an empty asset name, or, for assets to have identical asset names under different policy ids. 
+
+Because assets are manipulated in several user-facing features, and also via hardware devices, it is useful to come up with a short(er) and human-readable identifier for assets that user can recognize and refer to when talking about assets.
+
+# Specification
+
+We define a unique identifier for assets in pseudo-code as:
+
+```
+assetId := encodeBech32
+  ( datapart = hash
+    ( algorithm = 'blake2b'
+    , digest-length = 32
+    , message = policyId | assetName
+    )
+  , humanReadablePart = 'token'
+  )
+```
+
+where `|` designates the concatenation of two bytestrings and the `digest-length` is given in `bytes`.
+
+# Rationale
+
+- The asset id needs to be unique and must therefore include both the policy id and the asset name.
+- Using a hash gives us asset id of a same deterministic length which is short enough to display reasonably well on small screens.
+- The hash digest is short but long enough to still allow ~2^256 possible different asset ids, which is _plenty_.
+- We use bech32 as a user-facing encoding since it is both user-friendly and quite common within the Cardano eco-system (e.g. addresses, pool ids, keys).
+
+# Backwards Compatibility
+
+N/A
+
+# Reference Implementation
+
+<details>
+  <summary>Haskell</summary>
+
+```hs
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE TypeApplications #-}
+
+-- package: base >= 4.0.0
+import Prelude
+import Data.Function
+    ( (&) )
+
+-- package: bech32 >= 1.0.2
+import qualified Codec.Binary.Bech32 as Bech32
+
+-- package: bech32-th >= 1.0.2
+import Codec.Binary.Bech32.TH
+    ( humanReadablePart )
+
+-- package: bytestring >= 0.10.0.0
+import Data.ByteString
+    ( ByteString )
+
+-- package: cryptonite >= 0.22
+import Crypto.Hash
+    ( hash )
+import Crypto.Hash.Algorithms
+    ( Blake2b_256 )
+
+-- package: memory >= 0.14
+import Data.ByteArray
+    ( convert )
+
+-- package: text >= 1.0.0.0
+import Data.Text
+    ( Text )
+
+
+newtype PolicyId = PolicyId ByteString
+newtype AssetName = AssetName ByteString
+newtype AssetId = AssetId Text
+
+mkAssetId :: PolicyId -> AssetName -> AssetId
+mkAssetId (PolicyId pid) (AssetName name)
+    = (pid <> name)
+    & convert . hash @_ @Blake2b_256
+    & Bech32.encodeLenient hrp . Bech32.dataPartFromBytes
+    & AssetId
+  where
+    hrp = [humanReadablePart|token|]
+```
+</details>
+
```

Comment:
It would be great to have test vectors with actual policyId/asset names and the respective bech32 ids

Reply:
I agree!

Reply:
Done!